Enhance/create DRBD mdule and state

nick-wang opened the pull request:
### What does this PR do?
Enhance the DRBD module to not only monitor but also manage resource.
Also implement corresponding state for DRBD.

### What issues does this PR fix or reference?
As promised in previous [PR](https://github.com/saltstack/salt/pull/50410). Complete the functions for DRBD to manage the resource.
This also support multiple nodes with multiple volumes of DRBD resource.

All module/state functions are (unit/function) tested in the real DRBD environment.

### Previous Behavior
Able to monitor the DRBD resource status, but limited by 2 nodes with maximum 1 volume.

### New Behavior
Able to manage DRBD resource, support multiple nodes with multiple volumes.

### Tests written?

Yes. 100% coverage.

### Commits signed with GPG?

No

nick-wang committed:
Polish the drbd module overview and status functions.

Replace global variable with salt built-in __context__
Change self.assert to generic assert method.
Increase test coverage.
Polish the comments and __doc__
Remove dummy code for pylint.

nick-wang committed:
Add more drbd module functions to manage the latest DRBD.

Not only monitor DRBD, but also able to manage DRBD
like start/stop/promote/demote/etc...
Test coverage 100%

nick-wang committed:
Add drbd state and test cases.

DRBD state based on its salt module. Able to let resource
started/stopped/promoted/demoted/etc...
Test coverage 100%

nick-wang commented:
All codes are tested and run with DRBD environment. 100% unit test coverage.
Codes are follow the salt style, most issues reported in codeclimate don't make it better maintain/read.

Please help to review, thanks a lot in advanced.

nick-wang commented:
@cachedout @garethgreenaway Sorry for the trouble if i should not ping you directly.(I find you via the last PR review of DRBD module and the latest active maintainer of review.)

Could you help to review or find anyone suitable to help on review? Seems i don't have access to request team-core to review.
Thanks for the help.

Ch3LL committed:
Merge branch 'develop' into drbd-state-module

Ch3LL commented:
@nick-wang thanks for this PR! I have tagged it to be included in the neon release. I plan on creating the neon branch today and I will migrate this PR over when i do. I will review this either tomorrow or monday so we can get this PR progressing.

Ch3LL commented:
FYI I migrated this PR from develop to neon to ensure it is included in the upcoming neon release. Let me know if this caused any issues. Thanks

nick-wang commented:
@Ch3LL Thanks a lot for your great help! 
Please let me know if i need to do any change or update, thank.

Ch3LL commented on salt/modules/drbd.py:
```diff
@@ -1,13 +1,47 @@
-DRBD administration module
+Module to provide DRBD functionality to Salt
+
+.. versionadded:: pending
```
since this module is not new to neon can you remove the versionadded here and add `.. versionadded:: neon` to each of the new functions's doc string in the module?

nick-wang replied:
Add to all new function of module.

Ch3LL commented on salt/modules/drbd.py:
```diff
@@ -1,13 +1,47 @@
+import salt.utils.json
+import salt.utils.path
+
+LOGGER = logging.getLogger(__name__)
```
is there a reason you are changing this to `LOGGER` instead of `log`. It looks like in salt's code the convention is `log = loggin..`

nick-wang replied:
Will follow the salt style, change to `log`

Ch3LL commented on salt/states/drbd.py:
```diff
@@ -0,0 +1,462 @@
+#
+# Author: Nick Wang <nwang@suse.com>
+#
+# Copyright 2019 SUSE LINUX GmbH, Nuernberg, Germany.
```
can you remove the copyright? The license should not need to be here either as the apache license does not require it be added to each file.

nick-wang replied:
Removed the duplicate license in the new commit.

Ch3LL commented on salt/states/drbd.py:
```diff
@@ -0,0 +1,462 @@
+'''
+State module to provide DRBD functionality to Salt
+
+.. versionadded:: pending
```
```suggestion
.. versionadded:: neon
```

nick-wang replied:
Changed in the state file.

Ch3LL commented:
looks like the test `unit.test_doc.DocTestCase.test_states_doc_files` is failing because your new state module requires a rst file. you can look at the `doc/ref/states/all/` directory for examples. You will also need to add it to `doc/ref/states/all/index.rst`. If you could add a note to the neon release notes here ./doc/topics/releases/neon.rst as well about these new features that would be great. Thanks for this contribution. this will be great to have.

nick-wang committed:
Fix based on the review comments.

pull request: https://github.com/saltstack/salt/pull/52938
Add .. versionadded:: neon
Replace variable LOGGER to log.
Remove the duplicate apache license.

nick-wang committed:
Add drbd state doc file and update the neon release note.

nick-wang commented:
> looks like the test `unit.test_doc.DocTestCase.test_states_doc_files` is failing because your new state module requires a rst file. you can look at the `doc/ref/states/all/` directory for examples. You will also need to add it to `doc/ref/states/all/index.rst`. If you could add a note to the neon release notes here ./doc/topics/releases/neon.rst as well about these new features that would be great. Thanks for this contribution. this will be great to have.

Thanks for the opinions. Added the doc file and updated the neon release note in the new commits.

Ch3LL commented on salt/modules/drbd.py:
```diff
@@ -1,14 +1,50 @@
-DRBD administration module
+Module to provide DRBD functionality to Salt
+
+.. versionadded:: neon
```
just one last thing. if you could remove this line here since the module itself wasn't added in neon just the new functions. thanks

nick-wang replied:
Thanks! Removed in the new commit.
Sorry for the misunderstanding of the previous fix.

nick-wang committed:
Remove the wrong versionadd in drbd module.

drbd module is not completely new added, only
some new functions. Remove the first versionadded.

Ch3LL committed:
Merge branch 'neon' into drbd-state-module

nick-wang commented:
@Ch3LL @garethgreenaway Thanks for the review.

Ch3LL committed:
Merge branch 'neon' into drbd-state-module

dwoz committed:
Merge branch 'neon' into drbd-state-module

dwoz commented:
re-run full all

Ch3LL commented:
re-run full all

dmurphy18 reviewed (changes requested):
This is merging into saltstack:neon ? should that be master now.
doc/topics/releases/neon.rst, should that now be sodium.rst given the sodium release
Similarly the versionadded:: neon need to change to sodium
pylint needs to be run given there is use of {} in string handling to which pylint issues warnings

Ch3LL commented:
Following up here, looks like this PR needs to be migrated to the master branch as @dmurphy18 previously pointed out. Do you have the bandwidth to do so?
